app/controllers/HomeController.scala

```diff
@@ -60,107 +60,122 @@ class HomeController @Inject()(val homeView: views.html.Home,
         errorHandler(isAgent).showInternalServerError()
     }
-  }
 
-  def show(origin: Option[String] = None): Action[AnyContent] = auth.authenticatedAction(isAgent = false) {
-    implicit user =>
-      handleShowRequest(isAgent = false, origin)
+  private def buildHomePage(nextUpdatesDueDates: Seq[LocalDate], isAgent: Boolean)
+                           (implicit user: MtdItUser[_]): Future[Result] =
+    for {
+      unpaidCharges             <- financialDetailsService.getAllUnpaidFinancialDetails(isEnabled(CodingOut))
+      paymentsDue                = getDueDates(unpaidCharges)
+      dunningLockExists          = hasDunningLock(unpaidCharges)
+      outstandingChargesModel   <- getOutstandingChargesModel(unpaidCharges)
+      outstandingChargeDueDates  = getRelevantDates(outstandingChargesModel)
+      overDuePaymentsCount       = calculateOverduePaymentsCount(paymentsDue, outstandingChargesModel)
+      paymentsDueMerged          = mergePaymentsDue(paymentsDue, outstandingChargeDueDates)
+    } yield {
+
+      val nextUpdatesTileViewModel = NextUpdatesTileViewModel(nextUpdatesDueDates, dateService.getCurrentDate)
+
+      auditingService.extendedAudit(HomeAudit(user, paymentsDueMerged, overDuePaymentsCount, nextUpdatesTileViewModel))
+
+      val paymentCreditAndRefundHistoryTileViewModel =
+        PaymentCreditAndRefundHistoryTileViewModel(unpaidCharges, isEnabled(CreditsRefundsRepay), isEnabled(PaymentHistoryRefunds))
+
+      Ok(view(
+        isAgent = isAgent,
+        dunningLockExists = dunningLockExists,
+        nextPaymentDueDate = paymentsDueMerged,
+        currentTaxYear = dateService.getCurrentTaxYearEnd,
+        overDuePaymentsCount = Some(overDuePaymentsCount),
+        nextUpdatesTileViewModel = nextUpdatesTileViewModel,
+        displayCeaseAnIncome = user.incomeSources.hasOngoingBusinessOrPropertyIncome,
+        paymentCreditAndRefundHistoryTileViewModel = paymentCreditAndRefundHistoryTileViewModel
+      ))
+    }
+
+  private def getDueDates(unpaidCharges: List[FinancialDetailsResponseModel]): List[LocalDate] =
+    (unpaidCharges collect {
+      case fdm: FinancialDetailsModel => fdm.validChargesWithRemainingToPay.getAllDueDates
+    })
+      .flatten
+      .sortWith(_ isBefore _)
+      .sortBy(_.toEpochDay())
+
+  private def getOutstandingChargesModel(unpaidCharges: List[FinancialDetailsResponseModel])
+                                        (implicit user: MtdItUser[_]): Future[List[OutstandingChargeModel]] =
+    whatYouOweService.getWhatYouOweChargesList(
+      unpaidCharges,
+      isEnabled(CodingOut),
+      isEnabled(MFACreditsAndDebits)
+    ) map {
+      case WhatYouOweChargesList(_, _, Some(OutstandingChargesModel(outstandingCharges)), _) =>
+        outstandingCharges.filter(_.isBalancingChargeDebit)
+                          .filter(_.relevantDueDate.isDefined)
```

Comment:
Yes. That's likely to be the case. But if there's an efficiency argument to be made, I'd want to see performance data and know that once version is clearly much better than the other, and that it is in a time-critical portion of the code (which I strongly suspect this isn't as it isn't a complex processing area, or in multiple nested loops).

Moreover, I spotted (but didn't do) a further change that might improve this greatly. I think the outstandingChargesModel and the outstandingChargesDueDates values can be collapsed into by mapping to Due Dates instead of holding onto the OutstandingChargesModel. They are used around lines 112, 113, but these uses can be converted over and make much more sense when based on a list of due dates. This change would need to have those filters separate so one could be converted to map. I'll put a comment on the TD ticket for the splitting as tiles so that this can be looked at at then.